src/constraints/investment.jl

```diff
@@ -1,62 +1,131 @@
         end
     end
 
     # - Maximum (i.e., potential) investment limit for flows
-    for y in Y, (u, v) in Fi[y]
-        if graph[u, v].capacity > 0 && !ismissing(graph[u, v].investment_limit[y])
-            bound_value = _find_upper_bound(graph, y, u, v)
-            JuMP.set_upper_bound(flows_investment[y, (u, v)], bound_value)
+    constraint_indices = _create_flow_investment_indices(connection)
+    filtered_indices =
+        filter(row -> row.capacity > 0 && !ismissing(row.investment_limit), constraint_indices)
+    if !isempty(filtered_indices)
+        filtered_indices.bound_value .= _find_upper_bound(
+            filtered_indices.investment_limit,
+            filtered_indices.capacity,
+            filtered_indices.investment_integer,
+        )
+        for (flows_investment, row) in
+            zip(variables[:flows_investment].container, eachrow(filtered_indices))
+            JuMP.set_upper_bound(flows_investment, row.bound_value)
         end
     end
 end
 
-function _find_upper_bound(graph, year, investments...; is_bound_for_energy = false)
-    graph_investment = graph[investments...]
-    if !is_bound_for_energy
-        bound_value = graph_investment.investment_limit[year] / graph_investment.capacity
-        if graph_investment.investment_integer
-            bound_value = floor(bound_value)
-        end
-    else
-        bound_value =
-            graph_investment.investment_limit_storage_energy[year] /
-            graph_investment.capacity_storage_energy
-        if graph_investment.investment_integer_storage_energy
-            bound_value = floor(bound_value)
-        end
-    end
+function _create_asset_investment_indices(connection)
+    return DuckDB.query(
+        connection,
+        "SELECT
+            indices.asset,
+            indices.milestone_year,
+            indices.investment_integer,
+            asset.capacity,
+            asset_commission.investment_limit,
+        FROM indices_for_assets_investment AS indices
+        LEFT JOIN asset
+            ON asset.asset = indices.asset
+        LEFT JOIN asset_commission
+            ON asset_commission.asset = indices.asset
+                AND asset_commission.commission_year = indices.milestone_year
+        ",
+    ) |> DataFrame
+end
+
+function _create_asset_investment_energy_indices(connection)
+    return DuckDB.query(
+        connection,
+        "SELECT
+            indices.asset,
+            indices.milestone_year,
+            indices.investment_integer_storage_energy,
+            asset.capacity_storage_energy,
+            asset_commission.investment_limit_storage_energy,
+        FROM indices_for_assets_investment_energy AS indices
+        LEFT JOIN asset
+            ON asset.asset = indices.asset
+        LEFT JOIN asset_commission
+            ON asset_commission.asset = indices.asset
+                AND asset_commission.commission_year = indices.milestone_year
```

Comment:
Also add a `ORDER BY index` to make sure that follow the initial order to correctly zip with the variables.